cpp/src/copying/gather.cuh

```diff
@@ -0,0 +1,337 @@
+    const bool dest_has_nulls = destination_valid_col != nullptr;
+
+    const int warp_size = 32;
+
+    if (dest_has_nulls) {
+      size_type destination_row_base = blockIdx.x * blockDim.x;
+
+      size_type valid_count_accumulate = 0;
+
+      while (destination_row_base < num_destination_rows) {
+	size_type destination_row = destination_row_base + threadIdx.x;
+
+	const bool thread_active = destination_row < num_destination_rows;
+	size_type source_row =
+	  thread_active ? gather_map[destination_row] : 0;
+
+	const uint32_t active_threads =
+	  __ballot_sync(0xffffffff, thread_active);
+
+	bool source_bit_is_valid;
+	if (ignore_out_of_bounds) {
+	  if (0 <= source_row && source_row < num_source_rows) {
+	    source_bit_is_valid =
+	      src_has_nulls ? bit_mask::is_valid(source_valid_col, source_row)
+	      : true;
+	  } else {
+	    // If gather_map does not include this row we should just keep the
+	    // original value,
+	    source_bit_is_valid =
+	      bit_mask::is_valid(destination_valid_col, destination_row);
+	  }
+	} else {
+	  source_bit_is_valid =
+	    src_has_nulls ? bit_mask::is_valid(source_valid_col, source_row)
+	    : true;
+	}
+
+	// Use ballot to find all valid bits in this warp and create the output
+	// bitmask element
+	const uint32_t valid_warp =
+	  __ballot_sync(active_threads, source_bit_is_valid);
+
+	const size_type valid_index =
+	  cudf::util::detail::bit_container_index<bitmask_type>(
+	      destination_row);
+	// Only one thread writes output
+	if (0 == threadIdx.x % warp_size && thread_active) {
+	  destination_valid_col[valid_index] = valid_warp;
+	}
+	valid_count_accumulate += cudf::detail::single_lane_popc_block_reduce(valid_warp);
+
+	destination_row_base += blockDim.x * gridDim.x;
+      }
+      if (threadIdx.x == 0) {
+	atomicAdd(d_count + i, valid_count_accumulate);
+      }
+    }
+  }
+}
+
+
+/**---------------------------------------------------------------------------*
+ * @brief Function object for gathering a type-erased
+ * gdf_column. To be used with the cudf::type_dispatcher.
+ *
+ *---------------------------------------------------------------------------**/
+struct column_gatherer {
+  /**---------------------------------------------------------------------------*
+   * @brief Type-dispatched function to gather from one column to another based
+   * on a `gather_map`.
+   *
+   * @tparam column_type Dispatched type for the column being gathered
+   * @param source_column The column to gather from
+   * @param gather_map An iterator over integral values representing the gather map
+   * @param destination_column The column to gather into
+   * @param ignore_out_of_bounds Ignore values in `gather_map` that are
+   * out of bounds
+   * @param stream Optional CUDA stream on which to execute kernels
+   *---------------------------------------------------------------------------**/
+  template <typename column_type, typename iterator_type,
```

Comment:
`Element` is more descriptive as it's really the type of the _elements_ contained in the column rather than the type of the column.

```suggestion
  template <typename Element, typename MapIterator,
```